File: packages/js/src/platform/secrets.ts
import type { PlatformClient } from './client'

export interface Secret {
  key: string
  value: string
}

export interface SetSecretOpts {
  key: string
  value: string
  /** Target a specific service. Omit for project-scope. */
  serviceId?: string
  /** If true, set at project scope (shared across all services). */
  global?: boolean
}

export interface ListSecretsOpts {
  serviceId?: string
}

export class SecretsAPI {
  constructor(private readonly client: PlatformClient) {}

  /**
   * List secrets for a project or service.
   *
   * @param projectId - The project ID.
   * @param opts.serviceId - If given, list service-scoped secrets instead of project secrets.
   */
  list(projectId: string, opts?: ListSecretsOpts): Promise<Secret[]> {
    if (opts?.serviceId) {
      return this.client.get(`/api/apps/${opts.serviceId}/secrets`)
    }
    return this.client.get(`/api/projects/${projectId}/secrets`)
  }

  /**
   * Set one or more secrets.
   *
   * ```ts
   * // Service-scoped secret (default):
   * await lizard.secrets.set(projectId, { key: 'DATABASE_URL', value: '${{postgres.DATABASE_URL}}', serviceId })
   *
   * // Project-scoped (shared across all services):
   * await lizard.secrets.set(projectId, { key: 'LOG_LEVEL', value: 'info', global: true })
   * ```
   */
  async set(projectId: string, secrets: SetSecretOpts | SetSecretOpts[]): Promise<void> {
    const items = Array.isArray(secrets) ? secrets : [secrets]
    // Group by scope
    const byService = new Map<string, Record<string, string>>()
    const projectScoped: Record<string, string> = {}
    for (const s of items) {
      if (s.global || !s.serviceId) {
        projectScoped[s.key] = s.value
      } else {
        if (!byService.has(s.serviceId)) byService.set(s.serviceId, {})
        byService.get(s.serviceId)![s.key] = s.value
      }
    }
    const tasks: Promise<unknown>[] = []
    if (Object.keys(projectScoped).length > 0) {
      tasks.push(this.client.post(`/api/projects/${projectId}/secrets`, projectScoped))
    }
    for (const [serviceId, kv] of byService) {
      tasks.push(this.client.post(`/api/apps/${serviceId}/secrets`, kv))
    }
    await Promise.all(tasks)
  }

  /** Delete a secret by key. */
  async delete(projectId: string, opts: { key: string; serviceId?: string }): Promise<void> {
    if (opts.serviceId) {
      await this.client.delete(`/api/apps/${opts.serviceId}/secrets`, { key: opts.key })
    } else {
      await this.client.delete(`/api/projects/${projectId}/secrets`, { key: opts.key })
    }
  }
}

File: packages/python/lizard/platform/secrets.py
from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .client import PlatformClient


@dataclass
class Secret:
    key: str
    value: str | None
    service_id: str | None = None
    project_id: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "Secret":
        return cls(
            key=d["key"],
            value=d.get("value"),
            service_id=d.get("serviceId") or d.get("appId"),
            project_id=d.get("projectId"),
        )


class SecretsAPI:
    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def list(self, project_id: str, *, service_id: str | None = None) -> list[Secret]:
        """List secrets for a project (optionally filtered to a service)."""
        qs = f"?appId={service_id}" if service_id else ""
        result = self._client.get(f"/api/projects/{project_id}/secrets{qs}")
        return [Secret._from_dict(s) for s in (result if isinstance(result, list) else result.get("secrets", []))]

    def set(
        self,
        project_id: str,
        secrets: dict[str, str] | list[dict],
        *,
        service_id: str | None = None,
    ) -> None:
        """
        Set one or more secrets. Accepts a plain dict or a list of
        ``{"key": "K", "value": "V", "serviceId": "..."}`` objects.
        """
        if isinstance(secrets, dict):
            items = [{"key": k, "value": v, **({"appId": service_id} if service_id else {})}
                     for k, v in secrets.items()]
        else:
            items = [
                {**s, **({"appId": service_id} if service_id and "serviceId" not in s else {})}
                for s in secrets
            ]
        self._client.post(f"/api/projects/{project_id}/secrets", {"secrets": items})

    def delete(self, project_id: str, key: str, *, service_id: str | None = None) -> None:
        """Delete a secret by key."""
        body: dict = {"key": key}
        if service_id:
            body["appId"] = service_id
        self._client.delete(f"/api/projects/{project_id}/secrets", body)
